File: py/api/persona.py
OBSOLETE MOVED TO LOGIN#!/usr/bin/env python

"""
Handles the api call api/tumblrRequestToken

http://dev.imagediver.com/topic/image/cg/astoria_1923_0

http://neochronography.com/snap/astoria_1923_0/franklin.jpg

despite its name, this handles editing of snaps too

if the incoming data has a topic field, this is the snap being edited.

"""
from WebClasses import WebResponse,okResponse,failResponse

import ops.oauth_twitter as oauth_twitter
import store.dynamo as dynamo
import json
import misc
import constants
import store.models as models
import requests


verbose = True
testingFailure = False

def vprint(*args):
  if verbose:
    misc.printargs(args,"personaAPI")
    
    

    
def personaLogin(webin):
  
  cn = webin.content()
  #print "CONTENT",cn
  cob=json.loads(cn)
  assertion = cob["assertion"];
  vprint("assertion ",assertion);
  data = {'assertion':assertion,'audience':'http://dev.prototypejungle.org:80'}
  resp = requests.post('https://verifier.login.persona.org/verify', data=data, verify=True)
  vprint("ok",resp.ok)
  if resp.ok and (not testingFailure):
    # Parse the response
    verification_data = json.loads(resp.content)

    # Check if the assertion was valid
    vprint("verificationData",verification_data['status'])
    vprint("verificationData",verification_data);
    if verification_data['status'] == 'okay':
      email = verification_data['email']# Log the user in by setting a secure session cookie
      vprint("email",email)
      uname = "persona_"+email;    
      rs = {"userName":uname}
      usr = models.loadUserD(uname)
      if usr:
        vprint("ExisTing persona user",usr.__dict__)
        tac = getattr(usr,"accepted_terms",0)
        tac = 1 # no checking for terms yet
        handle = getattr(usr,"handle",None)
        if handle:
          rs["handle"] = handle
        rs["accepted_terms"] = tac
      else:
        vprint("New user")
        usr = models.UserD(uname)
        usr.accepted_terms = 1
        usr.save(True)
        rs["new_user"] = 1
      rs["accepted_terms"] = 1
      s = usr.newSession()
      sid = s.id
      rs["sessionId"] = sid
      return okResponse(rs) 
    else:
      return failResponse("notVerivied")
  else:
    return failResponse("failedRequestToPersona")

File: node/persona.js

var util = require('./util.js');

var dyno = require('./dynamo.js');

var pjdb = require('./db.js').pjdb;

var page = require('./page.js');

var session = require('./session');

var querystring = require('querystring');

var user = require('./user');

var https = require('https');

exports.login = function (request,response,cob) {
  util.log("persona","setHandleHandler",cob);
  function onVerifyResponse(vres) {
    util.log("persona","BACK FROM PERSONA");
    var data = "";
    vres.setEncoding('utf8');
    vres.on("data",function (chunk) {data+=chunk;});
    vres.on("end",function () {
      var verified = JSON.parse(data);
      util.log("persona","FROM PERSONA ",verified);
      if (verified.status == "okay") { // verified!
        var email = verified.email;
        util.log("persona","PERSONA USER ",email,"VERIFIED");
        var uname = "persona_T2_"+email;
        user.signIn(response,uname,true);   
      } else {
        page.servePage(response,"bad.html");
      }
    });
  }
  assertion = cob.assertion;
  util.log("persona","assertion ",assertion);
  audience = 'http://prototypejungle.org:8000'; // todo use a var dev vs non-dev
  util.log("persona","audience",audience);
  data = {'assertion':assertion,'audience':audience};
  util.log("persona","DATA",data);
  var datastring = querystring.stringify(data);
  var rq = https.request({'host':'verifier.login.persona.org',
                          'path':'/verify',
                          method:'POST'
                         },onVerifyResponse);
  rq.setHeader('Content-Type','application/x-www-form-urlencoded');
  rq.setHeader('Content-Length',datastring.length);
  util.log("persona","WRITING REQUEST");
  rq.write(datastring);
  rq.end();
}
        
                   